### ISSUE TITLE
r.KMvalues.filter_by_organism  seems to make to difference between KKM and KM

### ISSUE THREAD

**Issue by PolinaShichkova** [Robaina/BRENDApyrser#1]
Hello,

I have an impression that in some cases I got KKM in addition to KM while doing the following:

`brenda = BRENDA("brenda_download_01mar22.txt")
r = brenda.reactions.get_by_id('1.8.1.4')
tmp = []
tmp.append(r.KMvalues.filter_by_organism('Homo sapiens'))`

As a part of the output of the code above, I see the following:
`{'value': 4731.6,
    'species': ['Homo sapiens'],
    'meta': '#80# wild-type enzyme, pH 8.0, 37Â°C <135,136>',
    'refs': []},
   {'value': 4731.6,
    'species': ['Homo sapiens'],
    'meta': '#80# wild-type enzyme, pH 8.0, 37Â°C <135,136>',
    'refs': []},
   {'value': 3404.2,
    'species': ['Homo sapiens'],
    'meta': '#80# enzyme mutant P156A, pH 8.0, 37Â°C <135>',
    'refs': []},
   {'value': 1497.6,
    'species': ['Homo sapiens'],
    'meta': '#80# enzyme mutant P303A, pH 8.0, 37Â°C <135>',
    'refs': []}]}`

When I grep brenda_download_01mar22.txt for these numbers, e.g.
`grep "4731.6" brenda_download_01mar22.txt    `

I get:
**KKM**     #80# 4731.6 {NAD+}  (#80# wild-type enzyme, pH 8.0, 37°C <135,136>)
**KKM**     #80# 4731.6 {NAD+}  (#80# wild-type enzyme, pH 8.0, 37°C <135,136>)
        Biochim. Biophys. Acta (1973) 309, 289-295. {Pubmed:4731963} (c)
        (1973) 309, 307-317. {Pubmed:4731964}
        {Pubmed:4731965}
        tomatoes. Biochim. Biophys. Acta (1973) 309, 363-369. {Pubmed:4731966}
        (2002) 132, 935-943. {Pubmed:12473196}
        Mycoplasma pneumoniae. Protein J. (2008) 27, 303-308. {Pubmed:18473156}

and similarly only KKM on  
`grep "3404.2" brenda_download_01mar22.txt `
**KKM**     #80# 3404.2 {NAD+}  (#80# enzyme mutant P156A, pH 8.0, 37°C <135>) <135>
        Biochem. Parasitol. (1999) 99, 167-181. {Pubmed:10340482} (c)

and similarly only KKM on  
`grep "1497.6" brenda_download_01mar22.txt `

**KKM**     #80# 1497.6 {NAD+}  (#80# enzyme mutant P303A, pH 8.0, 37°C <135>) <135>
        Biofuels (2019) 12, 208. {Pubmed:31497068}
        Biofuels (2019) 12, 208. {Pubmed:31497068}
        Chem. Commun. (Camb. ) (2004) 5, 592-3. {Pubmed:14973623}
        (2013) 169, 77-87. {Pubmed:23149716}
        {Pubmed:23149756}
        37, 230-237. {Pubmed:14972646}
        J. Antimicrob. Agents (2011) 37, 585-587. {Pubmed:21497068}
        reticulum. Eur. J. Biochem. (1975) 51, 353-361. {Pubmed:1149736} (c)
        Biosci. (2004) 9, 1944-1953. {Pubmed:14977600}
        1016-1025. {Pubmed:11497462}
        54, 175-184. {Pubmed:1149746} (c)
        erythrocytes. Anal. Biochem. (1984) 141, 510-514. {Pubmed:6149706} (c)

Also, I noticed that if I comment out "[(BRENDA_KMs < 1000) & (BRENDA_KMs >= 0)]" of the example from Readme:
```
BRENDA_KMs = np.array([v for r in brenda.reactions
                       for v in r.KMvalues.get_values()])
values = BRENDA_KMs #[(BRENDA_KMs < 1000) & (BRENDA_KMs >= 0)]
import matplotlib.pyplot as plt

plt.hist(values)
plt.title(f'Median KM value: {np.median(values)}')
plt.xlabel('KM (mM)')
plt.show()
print(f'Minimum and maximum values in database: {values.min()} mM, {values.max()} mM')

```
Then min and max from Brenda txt file downloaded on 1st Mar 2022 are:
**Minimum and maximum values in database: -1995000000.0 mM, 2100000000.0 mM**
which don't look like correct Km mM values

Am I missing something or is it a bug?

Thank you in advance for addressing this issue

**Comment by Robaina** [comment]
Hi @PolinaShichkova ,

thanks for reporting this issue, by looking at the relevant part of the code:

https://github.com/Robaina/BRENDApyrser/blob/1b2f08f76040ff699f1c8876d9d30f1735f3fc83/brendapyrser/parser.py#L300

I think that your are right. The regex pattern in that case is picking up both "KM" and "KKM".

I'll fix it up and let you know,

Cheers

**Comment by Robaina** [comment]
So, I think the issue should be fixed by the latest commit. KMs should be of course non-negative since they are concentrations. The large negative values you were observing were due to the same problem (KKM + KM). However, I still get a minimum value of -999 for the set of all KM's in BRENDA. I checked this and this value is included in the database file. 

Perhaps is BRENDA's way to include an undefined KM value... For now it would make sense to restrict values to be non-negative...

**Comment by PolinaShichkova** [comment]
Thank you very much! I will test it out soon. Meanwhile, I am also working on some (less time-efficient than yours) regex based parser-script just to be double-sure (and to also extract some other stuff specific to my use case).

> For now it would make sense to restrict values to be non-negative...
-- Yes, I know these -999 are in the source file too, I'm filtering them out

Thanks a lot!

**Comment by Robaina** [comment]
Thank you!

**Comment by PolinaShichkova** [comment]
Hello,

I did a quick test and I think the issue has been fixed, but apparently it seems like there are still 101 Km entries with the value higher than 1000, even in the Brenda source file. So that I went to Brenda website to check one of the enzymes with such value (EC 1.1.1.1, https://www.brenda-enzymes.org/enzyme.php?ecno=1.1.1.1 ), and it was indeed even there! Then I checked the paper referenced in Brenda for this particular value (https://www.brenda-enzymes.org/literature.php?e=1.1.1.1&r=689887), and the units in the source publication didn't match Brenda's units! uM (micro) in source paper (https://link.springer.com/article/10.1007/s10930-008-9131-6/tables/1), mM in Brenda. I have created a Brenda-support ticket on their website for that.

Also just in case, if it may be useful for anyone in the future, I like testing regex stuff with https://regex101.com/

 Thank you!
